pkg/hashreader/hashreader.go

```diff
@@ -0,0 +1,120 @@
+package hashreader
+
+// HashReader writes what it reads from an io.Reader to an MD5 and SHA256 hash.Hash.
+import (
+	"bytes"
+	"crypto/md5"
+	"crypto/sha256"
+	"encoding/hex"
+	"hash"
+	"io"
+)
+
+// HashReader verifies that the content of the io.Reader matches the expected checksums.
+type HashReader struct {
+	md5sum, sha256sum []byte // Byte values of md5sum, sha256sum of client sent values.
+
+	src                 io.Reader
+	size                int64
+	md5Hash, sha256Hash hash.Hash
+}
+
+// New returns a new HashReader computing the MD5 sum and SHA256 sum
+// (if set) of the provided io.Reader.
+func New(src io.Reader, size int64, md5Hex, sha256Hex string) (*HashReader, error) {
```

Comment:
I think size can be omitted i.e. consumer needs to restrict by size using `io.LimitReader`

Reply:
Size is needed since we do not pass Size param anymore into our object interface.